### PR TITLE
Qiita2本目 M(マッチョ)VAEの作成

### LINKED ISSUE #101: [Qiita記事] CIFAR10の画像をVAEで生成してみた
## 執筆するQiita記事の概要
ワークアウトレポートでCIFAR10に対してVAEを適用しようとしたが力が足りず叶わなかった。
色々調べてみても中々良いのは出てこないので、様々な手段を用いてCIFAR10に対するVAEの性能を向上させる。

## 取り組む期間
~7/8
## 獲得を目指すスキル
- [x] CIFAR10にVAEを適用することでカラー画像を扱う際のノウハウを学ぶ
- [x] VAEに多く触れることでディープラーニングへの理解を深める
- [ ] VAEの潜在変数がどのように生成されているか理解する（エクストラ目標）
## Action list
- [x] CIFAR10をモノクロ加工してVAEできるかテストする
- [x] VAEの層数を変化させることで性能が上がるかテストする
- [x] VAEの潜在変数の次元数を変化させることで性能が上がるかテストする
- [ ] VAEがデータからどのように潜在変数を求めているか理解する（エクストラ目標）
- [ ] ミニレポート提出

### PR DESCRIPTION
**Pull request by takanoshuhei**
issue段階ではCIFAR10にVAEを適用する記事を書く予定だったが、マッチョに適応したほうがモチベーションの向上に加え真新しさもあったので内容を変更した。
マッチョのデータを集めるところから始め、VAEに適用し、結果の考察を行った。

Hackmdへのリンク(https://hackmd.io/@eq2bdyXdSY-1qr4ZXl4kOg/Hy8cGhKlH)

close #101  　

## 記事の目的
ありきたりなデータセットではない面白いVAEを作りたかった。
また、カラー画像に対して性能のいいVAEを作りたかったが、叶わなかった
## Action list
- VAEの原著の論文からVAEの問題設定をざっくりと理解する(完ぺきな理論の理解には知識が足りませんでした)
- 画像をweb上から集めるプログラムを作成した。
- 集めた画像をVAEに適用し、結果の考察を行った
- 色の情報が結果に関係しているのかを調べるために、白黒に加工した画像をVAEに適用し、考察した

## レビュワーへの質問
以下の点についてコメントして頂きたいです。
- VAEの原著の論文に挑んではみたのですが、予想以上に難しく、問題設定をきちんと理解できているか不安です。記事ではVAEについて、問題設定と何が難しいかだけざっくりと説明を行うので、そのざっくりな説明が間違っていないか教えていただきたいです。

- VAEを適用した結果があまり芳しくなく、精度を上げたいです。潜在変数の次元を上げれば精度が上がると考えているのですが、高次元の潜在空間を可視化するいい方法はないでしょうか。PCAで2次元まで次元削減をし、その座標から復元を行ってみたのですがそれでは良い結果は得られませんでした。
image
pcaを行った結果、同じ画像が復元されている。低次元から高次元へ写像してしまったことが原因？

- その他何か足りない点、分かりづらい説明、変な言葉遣いなどがあれば教えていただきたいです。

### CODE REVIEW

**Comment by FlowSs** [comment]
I am no expert in Image Recognition, but what I can tell from your code:

-The network layers are too small for your image size: you have 256x256 image, trained in unsupervised learning (no label) hence you have to build a deeper neural network for it to work well.
For instance, here an article (sorry in English) about supervised CNN trained on CIFAR-10 : https://appliedmachinelearning.blog/2018/03/24/achieving-90-accuracy-in-object-recognition-task-on-cifar-10-dataset-with-keras-convolutional-neural-networks/
The network is EXTREMELY complex and they have smaller pictures and label to help training, in your case you have to build something more complex to compensate since it's unsupervised.

-Don't forget, especially when involving CNN: Max Pooling, Dropout Layer and Batch Normalization can really increase learning accuracy !

-You indeed need to boost the latent space dimension: you cannot hope your VAE to learn to encoder 256x256 dimensions into a 2 dimensions. I don't think you need to know the latent space to generate your new data, just changing one or two latent variables at the same could work. If you really need to see what is happening, I recommend either using a SOM trained on your latent variables (can help you if your features are disentengled enough) or algorithm such as t-SNE to visualize on a 2D map your latent variable in higher dimension (careful though, depnding the dimension it can take a VERY long time to train, most people uses PCA to lower the dimension a bit then t-SNE). Straight using PCA won't work on such complex data, it's certainly highly non linear (and more complex than 2 dimension)

-Don't worry about getting blurred image, it's normal: One of the down side of VAE, is using the normal distribution as the approximation for the probability of the latent space. This causes this blurring effect (gaussian blur, that is used in photography !)

Hope it can help !

**Comment by takanoshuhei** [comment]
Hi! Mr.FrowSs! Thank you for the very polite comment!
> The network layers are too small for your image size: you have 256x256 image, trained in unsupervised learning (no label) hence you have to build a deeper neural network for it to work well.
The network is EXTREMELY complex and they have smaller pictures and label to help training, in your case you have to build something more complex to compensate since it's unsupervised.

Certainly the image size is quite large compared to CIFAR10 etc. I was not conscious of it. In order to build a deeper network, you seem to have to solve the memory problem as soon as possible.
For the first time I learned that having no labels makes the problem more difficult. However, since it is difficult to label this data set, I think that it can only cope with the network structure.

> Don't forget, especially when involving CNN: Max Pooling, Dropout Layer and Batch Normalization can really increase learning accuracy !

I'll try!

>You indeed need to boost the latent space dimension: you cannot hope your VAE to learn to encoder 256x256 dimensions into a 2 dimensions. I don't think you need to know the latent space to generate your new data, just changing one or two latent variables at the same could work. 

Is it better to recognize that if you raise the dimension of the latent space and move it by two dimensions in it, you can do the visualization you are currently doing? If it is, it is very thankful advice.

> Don't worry about getting blurred image, it's normal: One of the down side of VAE, is using the normal distribution as the approximation for the probability of the latent space. This causes this blurring effect (gaussian blur, that is used in photography !)

I felt a little relieved to hear that. Gauss blur will always occur. I don't care much about blurring, I would like to improve the model.

Thank you again for your polite advice! It was a great help!

**Review by TetraMiyazaki — APPROVED** [review]
お疲れ様です。基本的に高野くんとVAEは一緒にやってきたので、自分も同程度の理解です。

一つ気になったのが、モーフィングという単語でした。
https://www.learnopencv.com/face-morph-using-opencv-cpp-python/
image
モーフィング技法とは、IとJの画像のピクセルを上記のような関数の アルファを変えて  "滑らかな" モーフィング画像をつくっているようです。(モーフィングって **中間** 画像を作っている言葉から線形変換的な意味をもつ単語かと思ってましたが違うみたいですね。)

> ある画像から別の画像へ、連続的に変形させる画像表現

という意味合いで使っているのならいいんじゃないでしょうか。

引き続きラスト記事頑張ってください。

> PCAで2次元まで次元削減をし、その座標から復元を行ってみたのですがそれでは良い結果は得られませんでした。

これに対する考察が @senoura からあるそうです。

> その他何か足りない点、分かりづらい説明、変な言葉遣いなどがあれば教えていただきたいです。

記事の一番上に、「この記事は〜」の文言を入れてください。slackに詳細書いてます。

**Review by senoura — COMMENTED** [review]
大作お疲れ様です！記事の内容に対するコメントは追加でするものはありません。m(_ _)m
以下は質問についての考察です。
今回256*256ピクセルでデータの次元が60kくらいになっていますが実際に画像を表現するのに必要な次元はそんな膨大な数はいらないはずです。今回2次元に落としたようですが、潜在変数の次元は少なすぎても中途半端に多くてもダメだよーってことがこの記事で紹介されています。(GANの記事ですが(-_-;))
https://qiita.com/triwave33/items/a5b3007d31d28bc445c2
画像が変化するときの自由度なんかが関係しているのだと思いますが詳しいことは分からないのでドクターの方に聞いてみてください。(瀬野浦の指摘が的外れだったらごめんなさい(^人^))

**Comment by ae14watanabe** [comment]
cc: @kpasso1015 
image
このMとかIとかJが画素値なら`ディゾルブ`（変化前の画像がだんだん薄くなって、変化後の画像がだんだん濃くなるやつ）ってやつになりますね。つまり画像を線形補間してもあんまり面白くないんですよね。画像をコントロールしている潜在的なパラメータが1次元なら、直線ではなく曲線のような非線形な構造を持っているはずなので、Autoencoderでそれがうまく捉えられていれば綺麗に内挿できるでしょうって話ですね。

**Review by ae14watanabe — APPROVED** [review]
細かいコメントはHackMDに挙げています。要するにどこまで完成度を目指すのって話で、「VAEに自分の好きなやつ放り込んでみました！やっぱ実際は結構難しいですねー！」ぐらいであればもう十分だと思います。マッチョ面白いし。Approveします。

**Review by yoneda88 — COMMENTED** [review]
おつかれさま！記事の内容がおもしろかったです！！
Mustな修正は以下の１点だけです。
　「Variational Auto Encoder」って正式名称をどこかで一回は出そう！！できれば最初に
　登場以降は略称(AE or VAE)で良いけど。。。

それ以外の修正がMustじゃない細かいコメントはHackMDの方を見てください！

あと、手間ですけど投稿前に掲載したコードにエラーがないかの最終チェック(コピペして動かしてみる)はやってください！

**Review by senoura — APPROVED** [review]

**Comment by takanoshuhei** [comment]
@kpasso1015
返信遅れてすみません。コメントありがとうございます。
> > ある画像から別の画像へ、連続的に変形させる画像表現

> という意味合いで使っているのならいいんじゃないでしょうか。

自分はそういう意味合いで使っていました！それと「この記事は～」の文言も追加しました

**Comment by takanoshuhei** [comment]
> 大作お疲れ様です！記事の内容に対するコメントは追加でするものはありません。m(_ _)m 以下は質問についての考察です。 今回256*256ピクセルでデータの次元が60kくらいになっていますが実際に画像を表現するのに必要な次元はそんな膨大な数はいらないはずです。今回2次元に落としたようですが、潜在変数の次元は少なすぎても中途半端に多くてもダメだよーってことがこの記事で紹介されています。(GANの記事ですが(-_-;))
> https://qiita.com/triwave33/items/a5b3007d31d28bc445c2
> 画像が変化するときの自由度なんかが関係しているのだと思いますが詳しいことは分からないのでドクターの方に聞いてみてください。(瀬野浦の指摘が的外れだったらごめんなさい(^人^))

返信遅れてすみません。コメントありがとうございます。
やはりフロリアンさんも仰っていたように潜在変数の次元を2次元より多くしないといけなさそうですね。それを行うにはcolab上でメモリが足りずに動作できないという問題と、多次元の潜在変数をどう可視化するかがわからないという問題があるのでそこを解決しないといけないようです...

**Comment by takanoshuhei** [comment]
> 細かいコメントはHackMDに挙げています。要するにどこまで完成度を目指すのって話で、「VAEに自分の好きなやつ放り込んでみました！やっぱ実際は結構難しいですねー！」ぐらいであればもう十分だと思います。マッチョ面白いし。Approveします。

返信遅れてすみません。コメントありがとうございます。
今回の記事では期限的にも完成度にこだわるのは難しそうです...（本当はバズるくらいに完成度意識したいです)。
HackMDにコメントしてくださった指摘について、すぐに直せる部分は直しました！
図2の修正に関しては、現状zの次元が1になっているのでニューロンの数を増やして修正したいと思います。

**Comment by takanoshuhei** [comment]
> おつかれさま！記事の内容がおもしろかったです！！
> Mustな修正は以下の１点だけです。
> 　「Variational Auto Encoder」って正式名称をどこかで一回は出そう！！できれば最初に
> 　登場以降は略称(AE or VAE)で良いけど。。。
> 
> それ以外の修正がMustじゃない細かいコメントはHackMDの方を見てください！
> 
> あと、手間ですけど投稿前に掲載したコードにエラーがないかの最終チェック(コピペして動かしてみる)はやってください！

返信遅れてすみません。コメントありがとうございます。
Mustな修正であったVAEの正式名称の追加と、簡単に直せる部分については修正しました。
AEの説明と、VAEでなぜ確率分布を取り入れるのかそのお気持ちについてはすぐに修正することが難しそうです。今度質問させてください...

**Review by takuro-Ishida — APPROVED** [review]
お疲れ様です！
とてもチャレンジングで面白い記事だと思います.
自分が気になったところも基本的に、コメントにあったのでそれを修正していただければ問題ないかと思います.
完成度も@ae14watanabeがいってたようにVAEに好きな画像入れてみたというので今回は十分だと思いますのでApproveします！

**Review by Kusumoto-Keisuke — APPROVED** [review]
お疲れ様です．
読み物として，とても面白かったです．VAE正直分からなかったですが，楽しく読めました．
私は問題ないと思うのでApproveします．

**Comment by takanoshuhei** [comment]
VAEとAEの説明を追記しました！またさらにごちゃごちゃしちゃったので分かりづらい部分があればご教授ください！

**Review by yoneda88 — APPROVED** [review]
OKです！
VAEとAEの違いのところは、あんな感じでいいと思います！！注釈もついてるしw
